### DISCUSSION TITLE
How to send more data using the  function send_velocity_world_setpoint?

### DISCUSSION THREAD

**Discussion by Wallat1** [bitcraze/discussions#704] [Q&A]
Hi.

Could you please help me with the following problem?

I am using the function 
"def send_velocity_world_setpoint(self, vx, vy, vz, yawrate):", but now I want to add three more terms so in the file commander.py I have modified it as follows:

```
def send_velocity_world_setpoint(self, vx, vy, vz, yawrate, a, b, c):
         """
         Send Velocity in the world frame of reference setpoint with yawrate commands
         vx, vy, vz are in m/s
         yawrate is in degrees/s
         """
         pk = CRTPPacket()
         pk.port = CRTPPort.COMMANDER_GENERIC
         pk.channel = SET_SETPOINT_CHANNEL
         pk.data = struct.pack('<Bfffffff', TYPE_VELOCITY_WORLD,
                               vx, vy, vz, yawrate, a ,b, c)
         self._cf.send_packet(pk)
```

now as I understand it is also necessary to modify the file stabilizer_types.h adding these terms in

```
typedef struct setpoint_s {
   uint32_t timestamp;

   attitude_t attitude; // deg
   attitude_t attitudeRate; //deg/s
   quaternion_t attitudeQuaternion;
   float thrust;
   point_t position; // m
   velocity_t velocity; //m/s
   acc_t acceleration; // m/s^2
   bool velocity_body; // true if velocity is given in body frame; false if velocity is given in world frame

   structure {
     stab_mode_tx;
     stab_mode_t y;
     stab_mode_tz;
     stab_mode_troll;
     stab_mode_t pitch;
     stab_mode_t yaw;
     stab_mode_t quat;
   } mode;
} setpoint_t;
```

but from here on, I don't know what other file I have to modify. I hope you can help me.

Regards.

**Comment by krichardsson** [comment]
Hi, 

Set points are decoded in the Crazyflie in the `crtp_commander_generic.c` file. You can find a list of decoders that are connected to a specific type of set point, in your case the velocity world setpoint is decoded by the `velocityDecoder()` function. You could add your extensions here.

I assume your goal is to send a full state set point to the Crazyflie? There is actually an implementation of a full state setpoint in the Crazyflie already (it is used by Crazyswarm), but there is (not yet) a send function in the python lib. One option for you would be to add the send function in the python lib instead of extending the velocity world setpoint. 
The full state setpoint is decoded by the `fullStateDecoder()` function. The data structure is described in the doucmentation https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/crtp/crtp_generic_setpoint/#full-state
Note, the quaternion is transferred using a compressed format that there is also no support for in the python lib yet.

If you make a python implementation of the full state set point, please consider to contribute!

**Reply by knmcguire** [reply]
There was an attempt to merge the full state setpoint in the python lib here: https://github.com/bitcraze/crazyflie-lib-python/pull/382. But somehow the compression was not done right as it didn't work... but perhaps to take a look?

**Comment by Wallat1** [comment]
Hi.

Thank you very much for answering  @krichardsson and @knmcguire, I will check the information, if I get a good result I will make the contribution.

Regards.
